rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,212 @@
+  if (version < 7.8) {
+    const oldState = state as MyStateV77;
+    newState = { objectId: oldState.objectId, value: oldState.val };
+  } else {
+    newState = state;
+  }
+  
+  return newState;
+}
+
+const inject = (state: MyState, savedObjectReferences: SavedObjectReference[]) => {
+  return {
+    ...state,
+    objectId: savedObjectReferences.find(ref => ref.name = state.objectId)?.id;
+  }
+}
+
+const extract = (state: MyState) => {
+  const references = [{ name: 'objectId', id: state.objectId, type: 'savedObject' }];
+  return [{ ...state, objectId: 'objectId' }, references];
+}
+
+persistableStateService.register(MY_STATE_ID, { migrate, inject, extract });
+```
+
+```ts
+  const stateReadyForPersisting = persistableStateService.get(MY_STATE_ID),beforeSave(myState);
+  const stateReadyForConsuming = persistableStateService.get(MY_STATE_ID).afterLoad(stateReadyForPersisting);
+```
+
+# Motivation
+
+We need a formal way for authors of registry implementations to add migrations for this data. Since they may not own the saved object that their implementation ends up in, we can't rely on saved object migrations as it exists today (not to mention data that ends up in URLs that may need to be migrated).
+
+We also need to make sure that all the persited state containing references to saved objects extracts those before being saved and injects them later.
+
+# Detailed design
+
+```ts
+interface PersistableState extends Serializable {}
+
+interface PersistableStateDefinition {
+  id: string,
+  // migrate function receives state and version string and should return latest state version
+  // default is identity function
+  migrate: (state: PersistableState, version: string) => PersistableState,
+  // inject function receives state and a list of references and should return state with references injected
+  // default is identity function
+  inject: (state: PersistableState, references: SavedObjectReference[]) => PersistableState,
+  // extract function receives state and should return state with references extracted and array of references
+  // default returns same state with empty reference array
+  extract: (state: PersistableState) => [PersistableState, SavedObjectReference[]]
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
```

Comment:
Will this also run migrations? same question with `afterLoad`?

Reply:
afterLoad will run migrations (we don't know for how long state was stored and if its still latest version), beforeSave will not (we already give it the latest state, nothing to migrate)